### DISCUSSION TITLE
OpenMMLab 2.0公测 中文讨论区

### DISCUSSION THREAD

**Discussion by vansin** [open-mmlab/mmengine#526] [General]
# OpenMMLab 2.0 公测指南

OpenMMLab 2.0 于2022年9月1日在WAIC大会上发布，基于MMEngine通用、统一、灵活优雅的接口，重构了MMDetection、MMAction2 、MMClassification 、MMSegmentation 、MMDetection3D 、MMEditing 、MMPose 、MMTracking 、MMOCR 、MMGeneration 、MMRazor 、MMSelfSup和MMFlow算法库。

欢迎大家积极参与OpenMMLab 2.0的公测，有问题或者反馈直接写在评论区哇~

**温馨提醒**：OpenMMLab 2.0公测，请最好新建一个 conda 虚拟环境，防止和 OpenMMLab 1.0 已经安装的仓库冲突。

### 1.1. OpenMMLab 2.0 各算法库版本对应

|                      | OpenMMLab 1.0  | OpenMMLab 2.0                                                | OpenMMLab 2.0 公测 中文讨论分区                              |
| -------------------- | -------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| **MMEngine**         |                | [0.x](https://github.com/open-mmlab/mmengine)                |                                                              |
| **MMCV**             | 1.x            | [2.x](https://github.com/open-mmlab/mmcv/tree/2.x)           | [OpenMMLab 2.0公测中文讨论区之MMCV 2.x](https://github.com/open-mmlab/mmcv/discussions/2258) |
| **MMDetection**      | 0.x 、1.x、2.x | [3.x](https://github.com/open-mmlab/mmdetection/tree/3.x)    | [OpenMMLab 2.0公测中文讨论区之MMDetection 3.x](https://github.com/open-mmlab/mmdetection/discussions/8779) |
| **MMAction2**        | 0.x            | [1.x](https://github.com/open-mmlab/MMAction2/tree/1.x)      | [OpenMMLab 2.0公测中文讨论区之MMAction2 1.x](https://github.com/open-mmlab/mmaction2/discussions/1929) |
| **MMClassification** | 0.x            | [1.x](https://github.com/open-mmlab/MMClassification/tree/1.x) | [OpenMMLab 2.0公测中文讨论区之MMClassification 1.x](https://github.com/open-mmlab/mmclassification/discussions/1027) |
| **MMSegmentation**   | 0.x            | [1.x](https://github.com/open-mmlab/MMSegmentation/tree/1.x) | [OpenMMLab 2.0公测中文讨论区之MMSegmentation 1.x](https://github.com/open-mmlab/mmsegmentation/discussions/2064) |
| **MMDetection3D**    | 0.x            | [1.x](https://github.com/open-mmlab/MMDetection3D/tree/1.x)  | [OpenMMLab 2.0公测中文讨论区之MMDetection3D 1.1](https://github.com/open-mmlab/mmdetection3d/discussions/1829) |
| **MMEditing**        | 0.x            | [1.x](https://github.com/open-mmlab/MMEditing/tree/1.x)      | [OpenMMLab 2.0公测中文讨论区之MMEditing 1.x](https://github.com/open-mmlab/mmediting/discussions/1108) |
| **MMPose**           | 0.x            | [1.x](https://github.com/open-mmlab/MMPose/tree/1.x)         | [OpenMMLab 2.0公测中文讨论区之MMPose 1.x](https://github.com/open-mmlab/mmpose/discussions/1653) |
| **MMDeploy**         | 0.x            | [dev-1.x](https://github.com/open-mmlab/mmdeploy/tree/dev-1.x) | [OpenMMLab 2.0公测中文讨论区之MMDeploy dev-1.x](https://github.com/open-mmlab/mmdeploy/discussions/1035) |
| **MMTracking**       | 0.x            | [1.x](https://github.com/open-mmlab/MMTracking/tree/1.x)     | [OpenMMLab 2.0公测中文讨论区之MMTracking 1.x](https://github.com/open-mmlab/mmtracking/discussions/724) |
| **MMOCR**            | 0.x            | [1.x](https://github.com/open-mmlab/MMOCR/tree/1.x)          | [OpenMMLab 2.0公测中文讨论区之MMOCR 1.x](https://github.com/open-mmlab/mmocr/discussions/1386) |
| **MMRazor**          | 0.x            | [1.x](https://github.com/open-mmlab/MMRazor/tree/1.x)        | [OpenMMLab 2.0公测中文讨论区之MMRazor 1.x](https://github.com/open-mmlab/mmrazor/discussions/287) |
| **MMHuman3D**        | 0.x            |                                                              |                                                              |
| **MMSelfSup**        | 0.x            | [1.x](https://github.com/open-mmlab/MMSelfSup/tree/1.x)      | [OpenMMLab 2.0公测中文讨论区之MMSelfSup 1.x](https://github.com/open-mmlab/mmselfsup/discussions/481) |
| **MMFlow**           | 0.x            | [1.x](https://github.com/open-mmlab/MMFlow/tree/1.x)         |                                                              |
| **MMFewShot**        | 0.x            |                                                              |                                                              |

OpenMMLab 2.0 对应的版本也需要在readthedocs 中切换到对应的分支哟~

### 1.2. MMCV 和 MMEngine 的关系

如下表所示，MMCV 有两个大版本1.x 和 2.x，MMCV 1.x存在于OpenMMLab 1.0中，MMCV 2.x 和 MMEngine 存在于 OpenMMLab 2.0 中。

| OpenMMLab 1.0                                                | OpenMMLab 2.0                                                |
| ------------------------------------------------------------ | ------------------------------------------------------------ |
| MMCV 1.x 包含训练相关的模块（Runner、Hook等）和各种图像视频处理相关的模块（算子、io） | MMCV 2.x只包含Transform和自定义算子等和图像视频处理相关的模块 |
|                                                              | MMEngine 包含训练相关的模块（Registry、Runner、Hook等）      |

## 2. 贡献指南

[(1) pre-commit 约束代码规范](https://mmcv.readthedocs.io/zh_CN/latest/community/contributing.html)
[(2) 提PR指南](https://mmcv.readthedocs.io/zh_CN/latest/community/pr.html#)

**Comment by vansin** [comment]
大佬，问一下，mmengine的registery分级怎么继承mmdet，写上scope='mmdet'就继承了吗？比如让mmengnie当parent然后不同scope调用mmdet和mmseg该如何操作，谢谢[表情]

**Reply by vansin** [reply]
https://mmengine.readthedocs.io/zh_CN/latest/tutorials/registry.html#id3 可以参考这里~

**Comment by OpenMMLab-Coodinator** [comment]
https://mmengine.readthedocs.io/zh_CN/latest/tutorials/registry.html#id3 可以参考这里~

**Reply by vansin** [reply]
谢谢~

**Comment by vansin** [comment]
期待单元测试的教程~

**Comment by vansin** [comment]
项目真的是很好。但是说点不足可能更有价值。
这个问卷问哪里更好可能不太对吧。
现在感觉是基本文档和知识库不行，典型错误搜不到，一些错误处理看上去是没有自动测试透彻的结果。
比如算flops的，在某些算法没提供的话，他是直接报exception 而不是设计了兼容的提示错误。 
提供的工具比如转换，结果检查等新算法支持似乎并不统一。
这些其实对于转用这个库有很大影响，因为用的原因就是一个库什么都统一了，方便易用，现在每个算法的支持并不统一，体验就差了很多。
文档这么不完善，竟然出现在文件中是改进最多的项目。

**Reply by vansin** [reply]
Mmclassification 和 mmdetection 文档我都看过，其实我觉得资源还是比较好了，但是文档（就是正式的文档）相对一些比较好的项目还是差点。 基础文档是可以覆盖从头到尾的把项目跑起来的。 这个很重要。但是新版本似乎9-10月份就公布了，但是如果去新版本的文档，有一些还是空着的。举例来说，稳定版主线的文档目录就很全，但是3.x已经公布了一两个月了，文档里面内容很少，链接基本都是空的（basic concept)下面基本上就目录，没有内容。

**Reply by vansin** [reply]
你看这个截图可以看出来。 实际上，正式上手去用，主要以读官方文档为主，版本可以清晰区分，也够权威。 但是这里3.x 的文档严重不足。 如果有stable版本的经验还能稍微好点，如果没有基本上就太困难了。

**Reply by vansin** [reply]
您发的合集，我也看过，我觉得也挺不错的。但是那些其实都不如官方文档翔实些来的靠谱。我也看过一些官方的视频和讲解。我看到一些视频中会系统串一遍怎么跑起来，但是，他跟文档的作用肯定不一样。文档可以帮助工业化应用，如果 正式选型使用，文档的翔实程度就是前提，视频，教程都是零散知识，确实很难给一个团队说你们自己去看教程，遇到问题第一是查文档。效率和可靠性都在这里体现。 我觉得专业从业人员都能做到在文档翔实的情况下使用起来。 那些教程更多的感觉是为了帮人入门。 其实文档在入门方面可能效果更好

**Comment by vansin** [comment]
[from wechat talnex]
主要问题是继承混乱，这个然后方法定义不明确。原则上java那种接口形式的是最清晰的，但python弱化了这种概念，这导致mmdet3d开发出来的代码把继承特性仅仅当作reuse来实用，导致继承极其混乱，哪些函数需要重写，使用者都不清晰（不像pytorch Module子类很明确重新forward就好，autograd.Fuction明确重新forward和backward）实用者需要阅读源码，极大增加学习成本和维护成本 

先不说字典回溯困难。不论从任何角度，函数都应当仅接受自己操作的参数，字典这形式使得函数接收到本不属于自己的参数，使用者根本无法直观看出那些变量被修改，这种方式是很不安全的，而且*arg和**kwarg应该慎用，甚至不应当暴露给用户，这个参数同样会导致上述问题

我认为 mmdet3d并不是一个好的的 object-oriented 程序设计范例，虽然功能提供的很全面，但是感觉这种规模的团队不应该开发出这种学生级别的项目

**Comment by Cuppinono** [comment]
对着官方文档走下来，在训练的时候全是bug了，而且不易阅读

**Comment by whregrets** [comment]
在做mmyolov5实例分割模型导出onnx时，没有参考的demo，在使用mmdeploy里面实例分割的配置文件会在卡在InstanceData中检查value和self长度不一致这一步上，不知道该如何修改了，请问一下该如何解决这方面的问题，或者有没有相应的demo

**Comment by Broderick890** [comment]
请问这个项目下一阶段有什么开发计划?
